### PR TITLE
Select media device for video and audio input

### PR DESCRIPTION
**Pull request by Peterede**
The user should be able to select the microphone and video input device as well as the audio output device.  This pull request is for the issue https://github.com/nextcloud/spreed/issues/358

### CODE REVIEW

**Comment by fancycode** [comment]
Please rebase the merge request and remove unrelated commits from #1422 from this branch.

**Comment by Peterede** [comment]
rebased and removed unrelated commits

File: js/views/callinfoview.js [hunk 1]
```diff
@@ -91,6 +111,7 @@
 				canFullModerate: this._canFullModerate(),
 				isPublic: this.model.get('type') === 3,
 				showShareLink: !canModerate && this.model.get('type') === 3,
+				canPublish: OCA.SpreedMe.canPublish(),
```
Comment (nickvergessen) [inline js/views/callinfoview.js]: `TypeError: OCA.SpreedMe.canPublish is not a function`

File: js/app.js [hunk 2]
```diff
@@ -218,6 +218,106 @@
 			}.bind(this));
 		},
 
+		onAudioOutputChange: function() {
+			localStorage.setItem("audioOutput", $('#audioOutput').val());
```
Comment (nickvergessen) [inline js/app.js]:
```suggestion
			localStorage.setItem('audioOutput', $('#audioOutput').val());
```

File: js/app.js [hunk 3]
```diff
@@ -218,6 +218,106 @@
...
+
+		},
+
+		onAudioSourceChange: function() {
+			localStorage.setItem("audioSource", $('#audioSource').val());
```
Comment (nickvergessen) [inline js/app.js]:
```suggestion
			localStorage.setItem('audioSource', $('#audioSource').val());
```

Please use single quotes everywhere

File: js/app.js [hunk 4]
```diff
@@ -218,6 +218,106 @@
...
+
+			if (OCA.SpreedMe.app.signaling.currentCallToken !== null) {
+				this.changingMedia = true;
+				OCA.SpreedMe.app.signaling.leaveCurrentCall();
+				OCA.SpreedMe.webrtc.stopLocalVideo();
+
+				OCA.SpreedMe.webrtc.startLocalVideo(OCA.SpreedMe.webrtc.config.media);
+				OCA.SpreedMe.app.connection.joinCall(this.activeRoom.get('token'));
+
+					/*var senders = existingPeer.pc.getLocalStreams();
```
Comment (nickvergessen) [inline js/app.js]: If the code is not needed remove it.

File: js/views/callinfoview.js [hunk 5]
```diff
@@ -66,13 +66,33 @@
 		'		{{/if}}' +
 		'	</div>' +
 		'{{/if}}' +
-		'{{#if showShareLink}}' +
```
Comment (nickvergessen) [inline js/views/callinfoview.js]: Why was this removed?

File: js/views/callinfoview.js [hunk 6]
```diff
@@ -66,13 +66,33 @@
...
-		'	<div class="share-link-options">' +
-		'		<div class="clipboard-button"><span class="button icon-clippy"></span></div>' +
-		'	</div>' +
-		'{{/if}}' +
+		'<div class="talk-settings-button">' +
+		'		<span class="button icon-settings"></span>' +
+		'</div>' +
+		'</div>' +
```
Comment (nickvergessen) [inline js/views/callinfoview.js]:
```suggestion
```
invalid HTML structure

File: css/style.scss [hunk 7]
```diff
@@ -922,9 +922,66 @@ input[type="password"] {
...
+    linear-gradient(45deg, transparent 50%, gray 50%),
+    linear-gradient(135deg, gray 50%, transparent 50%);
+  background-position:
+    calc(100% - 20px) calc(1em + 2px),
+    calc(100% - 15px) calc(1em + 2px),
+    calc(100% - 2.5em) 0.5em;
+  background-size:
+    5px 5px,
+    5px 5px,
+    1px 1.5em;
```
Comment (nickvergessen) [inline css/style.scss]: this looks wrong

File: css/style.scss [hunk 8]
```diff
@@ -922,9 +922,66 @@ input[type="password"] {
...
 
+.icon-speaker {
+  background-image: url('../img/speaker.svg?v=1');
+  background-size: contain;
+}
+
+.settings-option {
+	background-position: left;
+	display: inline-block;
+    position: relative;
```
Comment (nickvergessen) [inline css/style.scss]: spaces vs. tabs

File: css/style.scss [hunk 9]
```diff
@@ -922,9 +922,66 @@ input[type="password"] {
...
+				&:focus,
+				&:active {
+					opacity: 1;
+				}
+			}
+		}
 	}
 }
 
+.icon-speaker {
```
Comment (nickvergessen) [inline css/style.scss]: instead do:
```scss
@include icon-black-white('menu-people', 'spreed', 1);
```

And you forgot to add the image.

File: css/style.scss [hunk 10]
```diff
@@ -922,9 +922,66 @@ input[type="password"] {
 				}
 			}
 		}
+		.talk-settings-button {
+			position: relative;
+			
+			.button {
+				cursor: pointer;
+				width: 50px;
```
Comment (nickvergessen) [inline css/style.scss]: should have the same size as all other icons

File: js/views/callinfoview.js [hunk 11]
```diff
@@ -66,13 +66,33 @@
...
+		'<div class="settings-menu hidden">' +
```
Comment (nickvergessen) [inline js/views/callinfoview.js]: I'm not sure this should be in the callinfo-section, as it is not specific to the conversation, but in general?

Maybe it's time for us to start using the settings feature in the left sidebar. However I'm not sure we should go through the trouble, when we can do this easily after switching to Vue.JS ( #1347 )

File: js/views/callinfoview.js (hunk 1 above)
Comment (Peterede) [reply js/views/callinfoview.js]: This is from another pull request https://github.com/nextcloud/spreed/pull/1298

File: js/views/callinfoview.js (hunk 5 above)
Comment (Peterede) [reply js/views/callinfoview.js]: This was not intentional, I have restored the removed section.

File: css/style.scss (hunk 7 above)
Comment (Peterede) [reply css/style.scss]: This is used for creating the downward facing arrow in the select list.  I have removed it and placed it elsewhere as it is not specifically needed.

File: css/style.scss (hunk 8 above)
Comment (Peterede) [reply css/style.scss]: The spaces have been replaced with tabs

File: css/style.scss (hunk 9 above)
Comment (Peterede) [reply css/style.scss]: The image has been added.

File: js/views/callinfoview.js (hunk 11 above)
Comment (Peterede) [reply js/views/callinfoview.js]: I think it is specific to the conversation just like the 'join call' button, the settings are used directly when joining the call and affect how the call appears and sounds.

**Comment by nickvergessen** [comment]
Closing as per https://github.com/nextcloud/spreed/pull/1450#discussion_r249744876

**Comment by djohle** [comment]
I'm curious what happened to the "as well as the audio output device" portion of this...?
